src/QsCompiler/CompilationManager/EditorSupport/CodeActions.cs

```diff
@@ -189,6 +230,50 @@ private static IEnumerable<TextEdit> OpenDirectiveSuggestions(this FileContentMa
                 .Select(edit => (edit.NewText.Trim().Trim(';'), file.GetWorkspaceEdit(edit)));
         }
 
+        /// <summary>
+        /// Returns a sequence of replacement Type suggestions for how errors for unknown types and callable in the given diagnostics can be fixed,
+        /// given the file for which those diagnostics were generated and the corresponding compilation.
+        /// Returns an empty enumerable if any of the given arguments is null.
+        /// </summary>
+        internal static IEnumerable<(string, WorkspaceEdit)> CapitalizationSuggestionsForUnknownIdentifiers
+            (this FileContentManager file, CompilationUnit compilation, IEnumerable<Diagnostic> diagnostics)
+        {
+            if (file == null || diagnostics == null) return Enumerable.Empty<(string, WorkspaceEdit)>();
+            var unknownCallables = diagnostics.Where(DiagnosticTools.ErrorType(ErrorCode.UnknownIdentifier));
+            var unknownTypes = diagnostics.Where(DiagnosticTools.ErrorType(ErrorCode.UnknownType));
+            if (!unknownCallables.Any() && !unknownTypes.Any()) return Enumerable.Empty<(string, WorkspaceEdit)>();
+
+            (string, WorkspaceEdit) SuggestedIdEdit(NonNullable<string> suggestedId, LSP.Range range)
+            {
+                var edit = new TextEdit { Range = range.Copy(), NewText = $"{suggestedId.Value}" };
+                return ($"Replace with \"{suggestedId.Value}\".", file.GetWorkspaceEdit(edit));
+            }
+
+            var suggestionsForIds = unknownCallables
+                .SelectMany(d => file.CapitalizationSuggestionsForIdAtPosition(d.Range.Start, compilation, out var _).Select(id => SuggestedIdEdit(id, d.Range)));
+            var suggestionsForTypes = unknownTypes
+                .SelectMany(d => file.CapitalizationSuggestionsForTypeAtPosition(d.Range.Start, compilation, out var _).Select(id => SuggestedIdEdit(id, d.Range)));
+            return suggestionsForIds.Concat(suggestionsForTypes);
+        }
+
+        /// <summary>
+        /// Returns a sequence of suggestions on how errors for unknown types and callable in the given diagnostics can be fixed, 
+        /// given the file for which those diagnostics were generated and the corresponding compilation. 
+        /// The given line number is used to determine the containing namespace. 
+        /// Returns an empty enumerable if any of the given arguments is null. 
+        /// </summary>
+        internal static IEnumerable<(string, WorkspaceEdit)> SuggestionsForUnknownIdentifiers
+            (this FileContentManager file, CompilationUnit compilation, int lineNr, IEnumerable<Diagnostic> diagnostics)
+        {
+            var suggestions = NamespaceSuggestionsForUnknownIdentifiers(file, compilation, lineNr, diagnostics);
+            if (!suggestions.Any())
+            {
+                suggestions = CapitalizationSuggestionsForUnknownIdentifiers(file, compilation, diagnostics);
```

Comment:
Since `lineNr` is not used here, it looks like the capitalization suggestions are for the entire file? When I tested it, VS Code only showed suggestions for the current line, so it might be doing some filtering, but it would probably be better to only send the relevant suggestions to the client.

Reply:
Nevermind, it looks like the list of diagnostics is provided by the code action context, so it shouldn't be for the whole file.